react-app/src/components/InputText/index.js

```diff
@@ -0,0 +1,87 @@
+import React, { PureComponent } from 'react';
+import PropTypes from 'prop-types';
+import FontAwesomeIcon from '@fortawesome/react-fontawesome';
+import { faEye, faEyeSlash } from '@fortawesome/fontawesome-free-solid';
+
+import styles from './index.css';
+
+export default class InputText extends PureComponent {
+  static TEXT = 'text';
+  static PASSWORD = 'password';
+
+  static propTypes = {
+    /**
+     * The type of the input. Possible values: 'text' or 'password'.
+     */
+    type: PropTypes.oneOf([InputText.TEXT, InputText.PASSWORD]).isRequired,
+
+    /**
+     * The label the input will display.
+     */
+    label: PropTypes.string,
+
+    /**
+     * The placeholder text the input will display.
+     */
+    placeholder: PropTypes.string,
+
+    /**
+     * The function to call when the input value changes.
+     */
+    onChange: PropTypes.func,
+
+    /**
+     * The function to call when enter is pressed.
+     */
+    onEnter: PropTypes.func,
+  };
+
+  static defaultProps = {
+    label: '',
+    placeholder: '',
+    onEnter: () => {},
+    onChange: () => {},
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = { isPasswordVisible: false };
+  }
+
+  _togglePasswordVisibility = () => {
+    this.setState({ isPasswordVisible: !this.state.isPasswordVisible });
+  }
+
+  _getInputIcon = () => {
+    if (this.props.type === InputText.TEXT) return null;
+
+    return (
+      <FontAwesomeIcon
+        className={styles.showHideButton}
+        icon={(this.state.isPasswordVisible) ? faEyeSlash : faEye}
+        onClick={this._togglePasswordVisibility}
+      />
+    )
+  }
+
+  render() {
+    const icon = this._getInputIcon();
+    const label = (this.props.label !== '') ? <label className={styles.label}>{this.props.label}</label> : null;
+    const placeholder = (this.props.placeholder !== '') ? this.props.placeholder : null;
+
+    return (
+      <div>
+        {label}
+        <input
+          className={(this.props.type === InputText.PASSWORD) ? styles.password : null}
+          type={(this.props.type === InputText.TEXT) ? InputText.TEXT :
```

Comment:
> from: https://en.wikiquote.org/wiki/Brian_Kernighan
> Everyone knows that debugging is twice as hard as writing a program in the first place. So if you're as clever as you can be when you write it, how will you ever debug it?

I think that by shrinking the code this way you are making it hard to figure out if something's wrong or if you can improve it somhow. And, if there's a bug it will be harder.

```jsx
type={(this.props.type === InputText.TEXT) ? InputText.TEXT :
      (this.state.isPasswordVisible) ? InputText.TEXT : InputText.PASSWORD}
```

Let's try to simplify it...

```jsx
let type;
if (this.props.type === InputText.TEXT) {
  type = InputText.TEXT;
} else {
  if (this.state.isPasswordVisible) {
    type = InputText.TEXT
  } else {
    type = InputText.PASSWORD;
  }
}
```

I've translated the double ternary operator into some ugly nested `if`. To me, this make it much more clear that there are some repeated cases that we could merge...

```jsx
let type;
if (this.props.type === InputText.TEXT || this.state.isPasswordVisible) {
  type = InputText.TEXT;
} else {
  type =  InputText.PASSWORD;
}
```

We can go a bit further and use the destructuring operator to shorten those lines...
Note that the destructuring usually takes place on the top of the function. This serves another purpose, you can quickly see what other data serves as "input" to your function.

```jsx
const { type } = this.props;
const { isPasswordVisible } = this.state;
let inputType;  // renamed variable to avoid collision
if (type === InputText.TEXT || isPasswordVisible) {
  inputType = InputText.TEXT;
} else {
  inputType =  InputText.PASSWORD;
}

// ...
<input type={inputType} />
```

Now that the code is simplified we could use it as a ternary operator and it will look simpler...

```jsx
const inputType = (type === InputText.TEXT || isPasswordVisible) ?  InputText.TEXT : InputText.PASSWORD;
```

Which looks simpler and (maybe) I wouldn't have argued to it on the first place.

The takeaway here is to write simpler code when possible, preferring longer/simpler code over complicated/shorter one.

Please use this simpler version of the code. Either the ternary operator one or the `if` based one. But extract the complexity from the prop passing.
I'd like the render method to be as clean as possible and avoid embedding logic on the prop passing whenever we can.

Reply:
Superb! Thank you for this review, it's just great and super clear!

IMHO, I prefer using using the props explicitly, I mean writing directly:
```const inputType = (this.props.type === InputText.TEXT || this.state.isPasswordVisible) ? InputText.TEXT : InputText.PASSWORD;```.
For me is more clear, is this OK for you? Is there any recommendation about this?

Reply:
Regarding using `this.props.whatever`, I think this is a matter of personal taste. The pro I see is that you can quickly see the data that affects the function. But, on the other hand it may not be direct that is a prop when you're looking at some code.
To me it's not incorrect either way, use whichever you prefer.